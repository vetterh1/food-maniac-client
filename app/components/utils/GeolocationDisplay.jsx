import React from 'react';
import PropTypes from 'prop-types';
import { Button, Modal, ModalHeader, ModalBody } from 'reactstrap';
import MdLocationOn from 'react-icons/lib/md/location-on';
import { connect } from 'react-redux';
import TestDisplayPositionFromStore from './TestDisplayPositionFromStore';


const styles = {

  locationOK: {
    color: 'green',
  },

  locationKO: {
    color: 'red',
  },
};


export class GeolocationDisplay extends React.Component {
  constructor() {
    super();
    this.toggle = this.toggle.bind(this);

    this.state = {
      statisticsOpen: false,
    };
  }

  toggle() {
    this.setState({ statisticsOpen: !this.state.statisticsOpen });
  }

  render = () => {
    return (
      <div>
        <Button onClick={this.toggle} size="md" color="link" style={this.props.coordinates.real ? styles.locationOK : styles.locationKO}>
          <MdLocationOn size={18} />
        </Button>
        <Modal isOpen={this.state.statisticsOpen} toggle={this.toggle}>
          <ModalHeader toggle={this.toggle}>Geolocation Statistics</ModalHeader>
          <ModalBody>
            <div className="geolocation_statistics">
              Coordinates:
              <ul>
                <li>Latitude: {this.props.coordinates.latitude ? Math.round(this.props.coordinates.latitude * 100000) / 100000 : 'unknown'}</li>
                <li>Longitude: {this.props.coordinates.longitude ? Math.round(this.props.coordinates.longitude * 100000) / 100000 : 'unknown'}</li>
                <li>Real: {this.props.coordinates.real ? 'true' : 'false'}</li>
                <li>Changed: {this.props.coordinates.changed ? 'true' : 'false'}</li>
                <li>Changed (real): {this.props.coordinates.changedReal ? 'true' : 'false'}</li>
              </ul>
            </div>

            <div className="geolocation_statistics">
              Statistics:
              <ul>
                <li>Nb refreshes: {this.props.coordinates.nbRefreshes}</li>
                <li>Nb different positions: {this.props.coordinates.nbDiffs}</li>
                <li>Nb real positions: {this.props.coordinates.nbReal}</li>
                <li>Nb estimated positions: {this.props.coordinates.nbEstimated}</li>
                <li>Nb closed positions: {this.props.coordinates.nbClose}</li>
              </ul>
            </div>

            <div className="geolocation_statistics">
              Simulation:
              <TestDisplayPositionFromStore onClick={this.toggle} />
            </div>
          </ModalBody>
        </Modal>
      </div>
    );
  }

}



GeolocationDisplay.propTypes = {
  coordinates: PropTypes.shape({
      latitude: PropTypes.number,
      longitude: PropTypes.number,
      real: PropTypes.boolean,
      changed: PropTypes.boolean,
      changedReal: PropTypes.boolean,
      nbRefreshes: PropTypes.number,
      nbDiffs: PropTypes.number,
      nbReal: PropTypes.number,
      nbEstimated: PropTypes.number,
      nbClose: PropTypes.number,
    }).isRequired,
};


// 1st to receive store changes
// Role of mapStateToProps: transform the "interesting" part of the store state
// into some props that will be received by componentWillReceiveProps
const mapStateToProps = (state) => { return { coordinates: state.coordinates }; };


export default connect(mapStateToProps)(GeolocationDisplay); // eslint-disable-line no-class-assign
