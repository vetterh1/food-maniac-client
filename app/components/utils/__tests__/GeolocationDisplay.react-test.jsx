
import React from 'react';
// import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';
import { Button, Modal, ModalHeader, ModalBody } from 'reactstrap';
import { GeolocationDisplay } from '../GeolocationDisplay';

describe('GeolocationDisplay', () => {
  // Minimal props to render component
  const minProps = {
    coordinates: {
      real: false,
    },
  };

  const realProps = {
    coordinates: {
      latitude: 25.1234,
      longitude: -10.4321,
      real: true,
    },
  };

  it('should render without throwing an error and match its empty snapshot', () => {
    const wrapper = shallow(<GeolocationDisplay {...minProps} />);
    expect(wrapper.length).toEqual(1);
    expect(wrapper).toMatchSnapshot();
  });

  it('should have a closed modal by default', () => {
    const wrapper = shallow(<GeolocationDisplay {...minProps} />);
    expect(wrapper.find(Modal).prop('isOpen')).toEqual(false);
  });

  it('should have a red button when not real', () => {
    const wrapper = shallow(<GeolocationDisplay {...minProps} />);
    expect(wrapper.find({ style: { color: 'red' } })).toHaveLength(1);
  });

  it('should have a green button when real', () => {
    const wrapper = shallow(<GeolocationDisplay {...realProps} />);
    expect(wrapper.find({ style: { color: 'green' } })).toHaveLength(1);
  });

  it('should open a modal on button click', () => {
    const wrapper = shallow(<GeolocationDisplay {...minProps} />);
    wrapper.find(Button).simulate('click');
    expect(wrapper.find(Modal).prop('isOpen')).toEqual(true);
  });
});
