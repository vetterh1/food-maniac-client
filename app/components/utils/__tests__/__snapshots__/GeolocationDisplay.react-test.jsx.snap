// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GeolocationDisplay should render without throwing an error and match its empty snapshot 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <div>
    <ei
        color="link"
        onClick={[Function]}
        size="md"
        style={
            Object {
                "color": "red",
              }
        }
        tag="button"
    >
        <MdLocationOn
            size={18}
        />
    </ei>
    <ei
        backdrop={true}
        isOpen={false}
        keyboard={true}
        toggle={[Function]}
        zIndex={1050}
    >
        <Ln
            tag="h4"
            toggle={[Function]}
            wrapTag="div"
        >
            Geolocation Statistics
        </Ln>
        <Rn
            tag="div"
        >
            <div
                className="geolocation_statistics"
            >
                Coordinates:
                <ul>
                    <li>
                        Latitude: 
                        unknown
                    </li>
                    <li>
                        Longitude: 
                        unknown
                    </li>
                    <li>
                        Real: 
                        false
                    </li>
                    <li>
                        Changed: 
                        false
                    </li>
                    <li>
                        Changed (real): 
                        false
                    </li>
                </ul>
            </div>
            <div
                className="geolocation_statistics"
            >
                Statistics:
                <ul>
                    <li>
                        Nb refreshes: 
                    </li>
                    <li>
                        Nb different positions: 
                    </li>
                    <li>
                        Nb real positions: 
                    </li>
                    <li>
                        Nb estimated positions: 
                    </li>
                    <li>
                        Nb closed positions: 
                    </li>
                </ul>
            </div>
            <div
                className="geolocation_statistics"
            >
                Simulation:
                <Connect(TestDisplayPositionFromStore)
                    onClick={[Function]}
                />
            </div>
        </Rn>
    </ei>
</div>,
  "nodes": Array [
    <div>
      <ei
            color="link"
            onClick={[Function]}
            size="md"
            style={
                  Object {
                        "color": "red",
                      }
            }
            tag="button"
      >
            <MdLocationOn
                  size={18}
            />
      </ei>
      <ei
            backdrop={true}
            isOpen={false}
            keyboard={true}
            toggle={[Function]}
            zIndex={1050}
      >
            <Ln
                  tag="h4"
                  toggle={[Function]}
                  wrapTag="div"
            >
                  Geolocation Statistics
            </Ln>
            <Rn
                  tag="div"
            >
                  <div
                        className="geolocation_statistics"
                  >
                        Coordinates:
                        <ul>
                              <li>
                                    Latitude: 
                                    unknown
                              </li>
                              <li>
                                    Longitude: 
                                    unknown
                              </li>
                              <li>
                                    Real: 
                                    false
                              </li>
                              <li>
                                    Changed: 
                                    false
                              </li>
                              <li>
                                    Changed (real): 
                                    false
                              </li>
                        </ul>
                  </div>
                  <div
                        className="geolocation_statistics"
                  >
                        Statistics:
                        <ul>
                              <li>
                                    Nb refreshes: 
                              </li>
                              <li>
                                    Nb different positions: 
                              </li>
                              <li>
                                    Nb real positions: 
                              </li>
                              <li>
                                    Nb estimated positions: 
                              </li>
                              <li>
                                    Nb closed positions: 
                              </li>
                        </ul>
                  </div>
                  <div
                        className="geolocation_statistics"
                  >
                        Simulation:
                        <Connect(TestDisplayPositionFromStore)
                              onClick={[Function]}
                        />
                  </div>
            </Rn>
      </ei>
</div>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 0,
      "_context": Object {},
      "_currentElement": <GeolocationDisplay
        coordinates={
                Object {
                        "real": false,
                      }
        }
/>,
      "_debugID": 1,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": GeolocationDisplay {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "props": Object {
          "coordinates": Object {
            "real": false,
          },
        },
        "refs": Object {},
        "render": [Function],
        "state": Object {
          "statisticsOpen": false,
        },
        "toggle": [Function],
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 1,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <div>
          <ei
                    color="link"
                    onClick={[Function]}
                    size="md"
                    style={
                              Object {
                                        "color": "red",
                                      }
                    }
                    tag="button"
          >
                    <MdLocationOn
                              size={18}
                    />
          </ei>
          <ei
                    backdrop={true}
                    isOpen={false}
                    keyboard={true}
                    toggle={[Function]}
                    zIndex={1050}
          >
                    <Ln
                              tag="h4"
                              toggle={[Function]}
                              wrapTag="div"
                    >
                              Geolocation Statistics
                    </Ln>
                    <Rn
                              tag="div"
                    >
                              <div
                                        className="geolocation_statistics"
                              >
                                        Coordinates:
                                        <ul>
                                                  <li>
                                                            Latitude: 
                                                            unknown
                                                  </li>
                                                  <li>
                                                            Longitude: 
                                                            unknown
                                                  </li>
                                                  <li>
                                                            Real: 
                                                            false
                                                  </li>
                                                  <li>
                                                            Changed: 
                                                            false
                                                  </li>
                                                  <li>
                                                            Changed (real): 
                                                            false
                                                  </li>
                                        </ul>
                              </div>
                              <div
                                        className="geolocation_statistics"
                              >
                                        Statistics:
                                        <ul>
                                                  <li>
                                                            Nb refreshes: 
                                                  </li>
                                                  <li>
                                                            Nb different positions: 
                                                  </li>
                                                  <li>
                                                            Nb real positions: 
                                                  </li>
                                                  <li>
                                                            Nb estimated positions: 
                                                  </li>
                                                  <li>
                                                            Nb closed positions: 
                                                  </li>
                                        </ul>
                              </div>
                              <div
                                        className="geolocation_statistics"
                              >
                                        Simulation:
                                        <Connect(TestDisplayPositionFromStore)
                                                  onClick={[Function]}
                                        />
                              </div>
                    </Rn>
          </ei>
</div>,
        "_debugID": 2,
        "_renderedOutput": <div>
          <ei
                    color="link"
                    onClick={[Function]}
                    size="md"
                    style={
                              Object {
                                        "color": "red",
                                      }
                    }
                    tag="button"
          >
                    <MdLocationOn
                              size={18}
                    />
          </ei>
          <ei
                    backdrop={true}
                    isOpen={false}
                    keyboard={true}
                    toggle={[Function]}
                    zIndex={1050}
          >
                    <Ln
                              tag="h4"
                              toggle={[Function]}
                              wrapTag="div"
                    >
                              Geolocation Statistics
                    </Ln>
                    <Rn
                              tag="div"
                    >
                              <div
                                        className="geolocation_statistics"
                              >
                                        Coordinates:
                                        <ul>
                                                  <li>
                                                            Latitude: 
                                                            unknown
                                                  </li>
                                                  <li>
                                                            Longitude: 
                                                            unknown
                                                  </li>
                                                  <li>
                                                            Real: 
                                                            false
                                                  </li>
                                                  <li>
                                                            Changed: 
                                                            false
                                                  </li>
                                                  <li>
                                                            Changed (real): 
                                                            false
                                                  </li>
                                        </ul>
                              </div>
                              <div
                                        className="geolocation_statistics"
                              >
                                        Statistics:
                                        <ul>
                                                  <li>
                                                            Nb refreshes: 
                                                  </li>
                                                  <li>
                                                            Nb different positions: 
                                                  </li>
                                                  <li>
                                                            Nb real positions: 
                                                  </li>
                                                  <li>
                                                            Nb estimated positions: 
                                                  </li>
                                                  <li>
                                                            Nb closed positions: 
                                                  </li>
                                        </ul>
                              </div>
                              <div
                                        className="geolocation_statistics"
                              >
                                        Simulation:
                                        <Connect(TestDisplayPositionFromStore)
                                                  onClick={[Function]}
                                        />
                              </div>
                    </Rn>
          </ei>
</div>,
      },
      "_renderedNodeType": 0,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <GeolocationDisplay
    coordinates={
        Object {
            "real": false,
          }
    }
/>,
}
`;
